servicenow/check.py

```diff
@@ -0,0 +1,199 @@
+            raise Exception('ServiceNow CMDB topology instance missing "basic_auth.user" value.')
+        if 'password' not in basic_auth:
+            raise Exception('ServiceNow CMDB topology instance missing "basic_auth.password" value.')
+
+        self.relation_types = {}
+
+        basic_auth_user = basic_auth['user']
+        basic_auth_password = basic_auth['password']
+        self.auth = (basic_auth_user, basic_auth_password)
+
+        self.base_url = instance['url']
+
+        self.instance_key = {
+            "type": self.INSTANCE_TYPE,
+            "url": self.base_url
+        }
+
+        self.instance_tags = instance.get('tags', [])
+
+        default_timeout = self.init_config.get('default_timeout', 5)
+        self.timeout = float(instance.get('timeout', default_timeout))
+
+
+        self._process_and_cache_relation_types()
+        self.start_snapshot(self.instance_key)
+        self._process_components()
+        self._process_component_relations()
+        self.stop_snapshot(self.instance_key)
+
+    def _collect_components(self):
+        """
+        collect components from ServiceNow CMDB's cmdb_ci table
+        :return: dict, raw response from CMDB
+        """
+        url = self.base_url + '/api/now/table/cmdb_ci?sysparm_fields=name,sys_id,sys_class_name,sys_created_on'
+
+        return self._get_json(url, self.timeout, self.auth)
+
+    def _process_components(self):
+        """
+        process components fetched from CMDB
+        :return: nothing
+        """
+        state = self._collect_components()
+
+        for component in state['result']:
+            id = component['sys_id']
+            type = {
+                "name": component['sys_class_name']
+            }
+            data = {
+                "name": component['name'].strip(),
+                "tags": self.instance_tags
+            }
+
+            self.component(self.instance_key, id, type, data)
+
+    def _collect_relation_types(self):
+        """
+        collects relations from CMDB
+        :return: dict, raw response from CMDB
+        """
+        url = self.base_url + '/api/now/table/cmdb_rel_type?sysparm_fields=sys_id,parent_descriptor'
+
+        return self._get_json(url, self.timeout, self.auth)
+
+    def _process_and_cache_relation_types(self):
+        """
+        collect available relations from cmdb_rel_ci and cache them in self.relation_types dict.
+        :return: nothing
+        """
+        state = self._collect_relation_types()
+
+        for relation in state['result']:
+            id = relation['sys_id']
+            parent_descriptor = relation['parent_descriptor']
+            self.relation_types[id] = parent_descriptor
+
+    def _collect_component_relations(self, offset, batch_size):
+        """
+        collect relations between components from cmdb_rel_ci and publish these in batches.
+        """
+        url = self.base_url + '/api/now/table/cmdb_rel_ci?sysparm_fields=parent,type,child'
```

Comment:
Could you add some links to servicenow docs you based this on? That will help future readers understand what is going on. Same for other places you use servicenow endpoints